src/emit/substrate/target.rs

```diff
@@ -967,46 +967,126 @@ impl<'a> TargetRuntime<'a> for SubstrateTarget {
         payload_len: IntValue<'b>,
         address: Option<PointerValue<'b>>,
         contract_args: ContractArgs<'b>,
-        _ty: ast::CallTy,
+        call_type: ast::CallTy,
         ns: &ast::Namespace,
         loc: Loc,
     ) {
         emit_context!(binary);
 
-        // balance is a u128
-        let value_ptr = binary
-            .builder
-            .build_alloca(binary.value_type(ns), "balance");
-        binary
-            .builder
-            .build_store(value_ptr, contract_args.value.unwrap());
-
         let (scratch_buf, scratch_len) = scratch_buf!();
-
         binary
             .builder
             .build_store(scratch_len, i32_const!(SCRATCH_SIZE as u64));
 
         // do the actual call
-        let ret = call!(
-            "seal_call",
-            &[
-                contract_args.flags.unwrap_or(i32_zero!()).into(),
-                address.unwrap().into(),
-                contract_args.gas.unwrap().into(),
-                value_ptr.into(),
-                payload.into(),
-                payload_len.into(),
-                scratch_buf.into(),
-                scratch_len.into(),
-            ]
-        )
-        .try_as_basic_value()
-        .left()
-        .unwrap()
-        .into_int_value();
+        let ret = match call_type {
+            ast::CallTy::Regular => {
+                // FIXME: Using the namespace value type, this comment is no longer valid?
+                // balance is a u128
+                let value_ptr = binary
+                    .builder
+                    .build_alloca(binary.value_type(ns), "balance");
+                binary
+                    .builder
+                    .build_store(value_ptr, contract_args.value.unwrap());
+                let ret = call!(
+                    "seal_call",
+                    &[
+                        contract_args.flags.unwrap_or(i32_zero!()).into(),
+                        address.unwrap().into(),
+                        contract_args.gas.unwrap().into(),
+                        value_ptr.into(),
+                        payload.into(),
+                        payload_len.into(),
+                        scratch_buf.into(),
+                        scratch_len.into(),
+                    ]
+                )
+                .try_as_basic_value()
+                .left()
+                .unwrap()
+                .into_int_value();
+                log_return_code(binary, "seal_call", ret);
+                ret
+            }
+            ast::CallTy::Delegate => {
+                // delegate_call asks for a code hash instead of an address
+                let hash_len = i32_const!(32); // FIXME: This is configurable like the address length
+                let code_hash_out_ptr = binary.builder.build_array_alloca(
+                    binary.context.i8_type(),
+                    hash_len,
+                    "code_hash_out_ptr",
+                );
```

Comment:
Can we use the scratch buffer for this and avoid the alloca?

Reply:
Having said that, alloca is cheap

Reply:
I like `alloca` better, see #1399